sdk/src/transaction/versioned/mod.rs

```diff
@@ -81,12 +81,23 @@ impl VersionedTransaction {
             Ordering::Equal => Ok(()),
         }?;
 
-        if signer_keys != expected_signer_keys {
-            return Err(SignerError::KeypairPubkeyMismatch);
-        }
-
         let message_data = message.serialize();
-        let signatures = keypairs.try_sign_message(&message_data)?;
+        let unordered_signatures = keypairs.try_sign_message(&message_data)?;
+        let signatures: Vec<Signature> = expected_signer_keys
+            .iter()
+            .map(|signer_key| {
+                let index = signer_keys
+                    .iter()
+                    .position(|key| key == signer_key)
+                    .ok_or(SignerError::KeypairPubkeyMismatch)?;
+
+                let signature = unordered_signatures
+                    .get(index)
+                    .ok_or_else(|| SignerError::InvalidInput("invalid keypairs".to_string()))?;
+
+                Ok(*signature)
+            })
+            .collect::<std::result::Result<_, SignerError>>()?;
```

Comment:
Signing is expensive compared to sorting. Since you know the expected signers and their order, you can sort the unordered signers against that. If any don’t match, you bail. Otherwise, then sign and you have the signatures in order.

Reply:
Great point. I've moved the keypair pubkey matching before the message signing. There's no way to reorder the `keypairs` themselves so I still need to reorder the signatures but at least this way we don't needlessly (or unexpectedly) sign when the keypairs are invalid

Reply:
Ah, makes sense. I was thinking of keypairs as like an array you could copy/sort, but it's just "thing with the `Signers` trait".